### PR TITLE
Allowed setters and getters to take additional arguments (superseded by #78991)

### LINKED ISSUE #3096: Improve access to variables in set / get properties GDScript 4.0
### Describe the project you are working on

I'm building an out of game, and in game room editor.  It has ships(like dungeons), decks (like floors), and rooms.  I want to be able to rapidly develop different ships, decks, and rooms, in the Godot Editor, these act as base designs/templates that can be edited, in game.

Within the room and deck layers I have GridMap based classes that do things like auto-tiling walls, and setting spaces for rooms to go in.  I've set this up so that in the editor, I can make a ship, add a deck to the ship, and a room to the deck.  Rooms have a set of export vars that determine how their walls and floors look, and work, and what the initial and current visibility states of those walls are.  However, many rooms will share the same settings, especially at the deck level(an engineering deck might have one look, while a bridge deck might have another).  And decks, likewise have the same variables as their rooms, as well as some other variables of their own.  These act as default values that the rooms can override, but that will initially not be set.  I plan to do the same with ships, providing default ship level variables that can be set at design time.

### Describe the problem or limitation you are having in your project

Whenever I set an exported variable on my room, the GridMap needs updated to display differently based on that new value.  This means I need to call an update function anytime the variable is set by the editor.  However, since I want to dynamically adjust these from the editor, at the deck(and eventually ship) level, I need to have each deck node, call an update to every room node, and each room node needs to then update itself.

In the end, I'll have at least 30 export vars at the room, deck, and ship level, each of which will basically look alike, here is an example with just 5:
```GDScript
@export var example_var:
	set(value):
		if is_nan(example_var) and is_nan(value):
			return
		if example_var == value:
			return
		example4_var = value
		update_child_params()
@export var example2_var:
	set(value):
		if is_nan(example2_var) and is_nan(value):
			return
		if example2_var == value:
			return
		example4_var = value
		update_child_params()
@export var example3_var:
	set(value):
		if is_nan(example3_var) and is_nan(value):
			return
		if example3_var == value:
			return
		example4_var = value
		update_child_params()
@export var example4_var:
	set(value):
		if is_nan(example4_var) and is_nan(value):
			return
		if example4_var == value:
			return
		example4_var = value
		update_child_params()
@export var example5_var:
	set(value):
		if is_nan(example5_var) and is_nan(value):
			return
		if example5_var == value:
			return
		example5_var = value
		update_child_params()
```
I have to manually(or regextually/macroedically) write the variable name 3 times, in what amounts to a simple 7 line function, for 30 variables.  It's not just annoying, it looks terrible to me.  Instead of a nice list of export vars, I've got this long list of code repeated over and over. 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

With just a little more access to the variable, itself, in the set property, setters would be so much more powerful, and allow a great deal of simplification, and reuse via functions.

Ideas for doing this(pick one):
- Change the set property from `set(value):` to `set(value, variable):`
Within the setter, `variable` would provide direct access to the variable, without re-calling the setter.  Use of ,variable would be optional.
- Make any references to the variable from functions called from the setter also have direct access to the variable.
In this case, it's a sub-scope, and could still allow boiling the setters down to 1 or 2 lines.
- Provide some way to set variables that skips the setter.
Some proposals:
  - Making `_variable` mean to give direct access to the variable(everywhere, not just in the setter).
  - Providing an engine level dictionary that gives direct access to every variable in memory(Lua, PHP, and I think Python all do something like this)
  - Provide a special function that will set a variable without calling the setter, such as `rawset(variable, value)` (Lua does this)
- Make anything that sets the variable, including the editor, call `_set()` so this can be handled in an override
- Provide a `super()` function for the setter that would perform `variable = value`
At some point in the setter, you pretty much have to do this function, anyway, unless setting it to something else, and you could just update value to something else, then call `super()`.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

With just a little more access to the variable, itself, in the set property, I could reduce this whole thing, potentially as small as the following:
```GDScript
@export var example_var1:
	set = handle_update_param
@export var example_var2:
	set = handle_update_param
@export var example_var3:
	set = handle_update_param
@export var example_var4:
	set = handle_update_param
@export var example_var5:
	set = handle_update_param
```
with the handle_update_param code looking like:
```GDScript
func handle_update_param(value, variable):
	if is_nan(variable) and is_nan(value):
		return
	if variable == value:
		return
		variable = value
	update_child_params()
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It's fair to say that similar end goals could be achieved without any one of these enhancements.  But for my scenario, even a couple of extra lines of script, turn in to 180 extra lines of very repetitive and readability impairing script.  I think a dict may often be considered as a solution to something like this, but it doesn't work out here, because a major goal is to be able to dynamically view the changes that adjusting these variables make in the editor, via the inspector, and switching to a dict, breaks the ability to use all of those inspector hints, etc.

### Is there a reason why this should be core and not an add-on in the asset library?

I think pretty much all of the solutions I propose dive in to the core of the GDScript code, and though I haven't yet taken a look at all of the core code involved in the solutions, I imagine several, including the top idea of just passing the variable to the set function, would easily be implemented in the core, do so without forcing users to adapt code to accommodate the changes, and come with fairly low risk.

### LINKED ISSUE #6750: Automate or reduce the boilerplate when forwarding script properties in GDScript
### Describe the project you are working on

A jrpg with stats-heavy combat and visual novel elements.

### Describe the problem or limitation you are having in your project

I often find myself having to write a lot of tedious boilerplate code to gracefully navigate through complicated objects. 

As a simple example, suppose I have a Character script that contains, among its many variables, a Stats resource (which defines attributes like hp, attack, defense, etc). One way to access the hp of a character would be `character.stats.hp`. However, best practices (eg: The Law of Demeter) dictate that it would be better to write `character.get_hp()` or simply `character.hp` so that the Character script doesn't need to know anything about the internal structure of the Stats resource. The most straightforward approach I can think of would be to write the Character script like so:

```
var stats : Stats
var hp : int :
    get: return stats.hp
    set(val): stats.hp = val
var defense : int :
    get: return stats.defense
    set(val): stats.defense = val
```
and so on for all the attributes I want Character to have direct access to. My problems with this approach are that

1. The code is tedious to write and feels like it could be automated easily
2. It adds a lot of visual noise to the beginning of a script file that makes it difficult to parse what variables a script has access to

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I see two approaches for overcoming this problem.

1. Adding some syntactic sugar (possibly by way of custom annotations or a new built-in annotation) that stands in for the boilerplate.
2. Let the editor auto-generate the boilerplate, perhaps via a custom plug-in or a new built-in editor feature.

### Syntactic Sugar Approach
Something along the lines of
```
var stats : Stats
@forward_to(stats) var hp : int
@forward_to(stats) var defense : int
```
@forward_to could be a new built-in annotation or a custom annotation if #1316 is ever implemented.

### Auto-generation Approach
When the user writes
```
var stats: Stats
```
there is some UI within the editor which allows the user to auto-generate code such as
```
# auto-generated setters and getters --------------------
func get_hp() -> int:
    return stats.hp


func set_hp(value : int) -> void:
    stats.hp = value
```
This code could be inserted towards the end of the script to avoid cluttering the script's header with visual noise.

The UI for achieving this could either be built in to the editor or provided by custom tooling/a plug-in.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

While a single property can be forwarded in 2-3 lines of code using getters/setters, these lines quickly add up over many properties. 

### Is there a reason why this should be core and not an add-on in the asset library?

Better built-in support for forwarding encourage the use of good design principles like Composition over Inheritance, Single Responsibility, and the Façade pattern. 

The Character script in my example respects these principles because it is composed of a Stats object (among other variables), instead of inheriting (from something like CharacterWithoutStats or EntityWithStats) or directly including the members of Stats (which could make Character have multiple responsibilities).

### LINKED ISSUE #74144: Setters/Getters don't allow functions with default arguments
### Godot version

4.0-rc6_win64

### System information

Windows 10

### Issue description

I was converting a Godot 3.5 project of mine to Godot 4 and found the following code throws an error:
```gdscript
var dims : set = set_dims

func set_dims(_dims, some_arg=null):
{
    print(some_arg)
    dims = _dims
}
```
Results in:

> Error: Function "set_dims" cannot be used as setter because of its signature.

I can luckily get the functionality to work with alternative syntax, but it's a bit silly and overly verbose when you have a lot of setters that should otherwise be very simple. It impacts readability quite a bit.

For example, the following:
```gdscript
var dims : set = set_dims
var surface : set = set_surface
var counts : set = set_counts
var bedrock : set = set_bedrock
var size : set = set_size
```
Becomes this:
```gdscript
var dims :
    set(_dims): set_dims(_dims)
var surface :
    set(_surface): set_surface(_surface)
var counts :
    set(_counts): set_counts(_counts)
var bedrock :
    set(_bedrock): set_bedrock(_bedrock)
var size :
    set(_size): set_size(_size)
```

I can't see any particular reason for this not to work, seeing as normal function calls can call functions without overriding default values. I'd probably call this a bug or a missing feature.

I found an issue back from 3.1 beta10, which seems to be when the original functionality was implemented: https://github.com/godotengine/godot/issues/26530

### Steps to reproduce

The sample code from above should do it:
```gdscript
var dims : set = set_dims

func set_dims(_dims, some_arg=null):
{
    print(some_arg)
    dims = _dims
}
```

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by nlupugla**
Closes https://github.com/godotengine/godot-proposals/issues/6750.
Closes https://github.com/godotengine/godot-proposals/issues/3096.
Closes https://github.com/godotengine/godot-proposals/discussions/6107 (discussion).
Closes https://github.com/godotengine/godot/issues/74144.

addresses the following proposals:
https://github.com/godotengine/godot-proposals/issues/2817.
https://github.com/godotengine/godot-proposals/issues/2983.
https://github.com/godotengine/godot-proposals/issues/4867.
https://github.com/godotengine/godot-proposals/issues/4868.

For example, the following is now possible:

```GDScript
@tool
extends Node2D

@onready var sprite := $Icon as Sprite2D
@export var offset : Vector2 : set=set_sprite_property, get=get_sprite_property
@export var centered : bool : set=set_sprite_property, get=get_sprite_property


func get_sprite_property(property : StringName) -> Variant:
	return sprite.get(property)
	
	
func set_sprite_property(property : StringName, value : Variant) -> void:
	sprite.set(property, value)
```

Previously, one would have had to use a more verbose and tedious approach:

```GDScript
@tool
extends Node2D

@onready var sprite := $Icon as Sprite2D
@export var offset : Vector2 : set=set_sprite_offset, get=get_sprite_offset
@export var centered : bool : set=set_sprite_centered, get=get_sprite_centered


func get_sprite_offset() -> Variant:
	return sprite.get("offset")
	
	
func set_sprite_offset(value : Variant) -> void:
	sprite.set("offset", value)


func get_sprite_centered() -> Variant:
	return sprite.get("centered")
	
	
func set_sprite_property(value : Variant) -> void:
	sprite.set("centered", value)
```

TODO: 

- [x] Attach some GDScript tests.
- [x] Require that the first argument of a 2-argument setter and a 1-argument getter be a StringName

### CODE REVIEW

File: modules/gdscript/gdscript_compiler.cpp
Comment (vnen) [inline modules/gdscript/gdscript_compiler.cpp:274]: This is getting the value from the wrong vector. Should use the static variables one:

```suggestion
					if (codegen.script->static_variables_indices[identifier].getter_size == 1) {
```

Comment (nlupugla) [reply modules/gdscript/gdscript_compiler.cpp:274]: Good catch! I amended my commit with the fix.

**Comment by nlupugla** [comment]
I noticed the tests complained about "name" shadowing a class variable. I wrote a few lines like `Variant name = p_name;`. I changed them to `Variant property_name = name`, so hopefully that fixes the issue.

**Review by dalexeev — CHANGES_REQUESTED** [review]
I tested and reviewed this PR and added some changes: signature analysis, support for optional arguments (godotengine/godot-proposals#6107, [vnen's comment](https://github.com/godotengine/godot-proposals/issues/6750#issuecomment-1528823707)), tests.

Commit: https://github.com/dalexeev/godot/commit/19b170afc525d3a3b19e9f41c3f4aff38ff8c8b2
Patch: [review.patch.txt](https://github.com/godotengine/godot/files/11451819/review.patch.txt)

You can apply the patch with `git apply` or `git am` (and then you need [to squash](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html) the commits into one). You can credit me as a co-author if you wish.

**Comment by nlupugla** [comment]
Thanks for the review @dalexeev :)
I applied your patch and pushed the changes, but I don't think I quite figured out the whole rebasing thing. I tried to squash my commits together and credit you as a coauthor, but got confused somewhere along the way...

File: modules/gdscript/tests/scripts/runtime/features/static_variables.out
Comment (dalexeev) [inline modules/gdscript/tests/scripts/runtime/features/static_variables.out:8]: This is a bug. This is why I prefer not to use `--gdscript-generate-tests` but to edit `.out` manually, even if it's tedious.

File: modules/gdscript/tests/scripts/runtime/features/setter_getter_with_property_argument.gd
Comment (dalexeev) [inline modules/gdscript/tests/scripts/runtime/features/setter_getter_with_property_argument.gd:29]: This file does not follow the GDScript style guide (wrong spaces, symbolic logical operators, missing trailing newline). This is undesirable in non-parser tests. In addition, I do not see the point in this test, in my opinion it duplicates the ones I have already added.

**Review by dalexeev — CHANGES_REQUESTED** [review]
There should be one commit. Use rebase, don't merge master into your branch.

File: modules/gdscript/tests/scripts/runtime/features/setter_getter_with_property_argument.gd
Comment (nlupugla) [reply modules/gdscript/tests/scripts/runtime/features/setter_getter_with_property_argument.gd:29]: Ah, I was going to ask you about this test on Rocket. I had written this a week or two ago before your patch. I noticed it was redundant with your tests. I can remove it.

**Comment by dalexeev** [comment]
I have prepared a squashed commit for you.

[squashed-commit.patch.txt](https://github.com/godotengine/godot/files/11460060/squashed-commit.patch.txt)

```shell
# upstream is godotengine/godot, origin is nlupugla/godot.
git fetch upstream
git checkout set-get-variable-property
git reset --hard upstream/master
git am squashed-commit.patch.txt
rm squashed-commit.patch.txt
git push --force origin
```

**Comment by nlupugla** [comment]
Thanks for preparing the squashed commit! Unless there's a rush to get the PR accepted, I want to see if I can figure out how to do it on my own, just for self-education. If there is a rush, let me know and I'll go ahead and apply the patch :)

**Comment by nlupugla** [comment]
Alright, I think I managed to figure out how to squash the commits and credit you. Thanks for the help :)
Just so you know, I haven't had a chance yet to test the changes you made. Should I trust that your changes are good and remove my "draft" status or would you appreciate if I did some testing?

**Comment by dalexeev** [comment]
> Alright, I think I managed to figure out how to squash the commits and credit you.

No, it doesn't look correct. You just rolled back to your old version and added my tests, but the rest of the changes were discarded (signature analysis, support for optional arguments, etc.). Also, `Co-authored-by` is incorrect.

> Should I trust that your changes are good and remove my "draft" status or would you appreciate if I did some testing?

I ran the tests locally, on my PC they complete successfully. Also I tested manually, looks correct.

Sorry for making the review in the form of an "ultimatum patch", I respect your desire to figure it out on your own. I just want to help you and save time, yours and mine. In any case, I want to say that as a first contribution, your work is already impressive!

**Comment by nlupugla** [comment]
Thanks :)
I think I fixed it now.

**Comment by dalexeev** [comment]
Yes, it looks correct now. :+1: Except the commit message could be improved. I suggested the following in the patch:

```
Improve variable setters and getters

* Add support for name argument.
* Allow optional arguments.
```

You can use something else if you feel like it.

Also, `Co-authored-by` usually uses the real name if known (not the nickname). But it doesn't matter, GitHub recognized me anyway. :)

**Comment by nlupugla** [comment]
Nice. I'm recompiling on my machine just to test things out again. When I'm satisfied, I'll remove my draft status.

File: modules/gdscript/gdscript_analyzer.cpp
Comment (dalexeev) [inline modules/gdscript/gdscript_analyzer.cpp:1337]:
```suggestion
										has_valid_setter = false;
```

My fault.

File: modules/gdscript/gdscript.cpp
Comment (dalexeev) [inline modules/gdscript/gdscript.cpp:941]: I just wanted to be consistent, but there are also `err` and `call_err` variables in this file. Not sure which is better.

Comment (nlupugla) [reply modules/gdscript/gdscript.cpp:941]: Fixed it :)

Comment (dalexeev) [reply modules/gdscript/gdscript.cpp:941]: You probably meant `has_valid_getter` copy-paste mistake.

File: modules/gdscript/gdscript_analyzer.cpp
Comment (dalexeev) [inline modules/gdscript/gdscript_analyzer.cpp:1326]:
```suggestion
						if (setter_function->mandatory_parameter_count == 0 || setter_function->mandatory_parameter_count > 2) {
```

File: modules/gdscript/gdscript.cpp
Comment (nlupugla) [reply modules/gdscript/gdscript.cpp:941]: Wait, I'm confused now. Is line 1337 (lol) supposed to set has_valid_setter or has_valid_getter? I figured it should be has_valid_setter because the nearby code is about setters.

Comment (dalexeev) [reply modules/gdscript/gdscript.cpp:941]: This comment is **not** on line 1337 (it's above, marked as "Outdated" and you've corrected it). This comment is about naming `CallError` variables (`err` vs `ce` vs `call_err`).

File: modules/gdscript/gdscript_analyzer.cpp
Comment (nlupugla) [reply modules/gdscript/gdscript_analyzer.cpp]: Fixed it.
Thanks for the help and patience!

**Comment by nlupugla** [comment]
I did some brief testing on my end and everything seemed to check out. I removed the draft status from my PR.

I edited my original post so that it mentions being able to have optional arguments in the setters and getters. Should I edit the title of my PR as well, or will that mess something up?

**Comment by dalexeev** [comment]
> Should I edit the title of my PR as well, or will that mess something up?

You can change the title to something more suitable, it will not change anything except the subject of email notifications.

Also please add this to your post:

Closes godotengine/godot-proposals#6750.
Closes godotengine/godot-proposals#3096.
Closes godotengine/godot-proposals#6107 (discussion).
Closes #74144.

Also this probably resolves or provides an alternative for the following proposals (we can not close them, but just write if/when this PR will be merged):

godotengine/godot-proposals#2817.
godotengine/godot-proposals#2983.
godotengine/godot-proposals#4867.
godotengine/godot-proposals#4868.

**Comment by dalexeev** [comment]
No, you need to repeat "Closes".

**Review by dalexeev — APPROVED** [review]
Looks good to me! There is only a minor suggestion for naming `CallError` variables. I think this PR is ready for review by the GDScript team.

**Comment by nlupugla** [comment]
> Looks good to me! There is only a minor suggestion for naming `CallError` variables. I think this PR is ready for review by the GDScript team.

Awesome :)

So should I name the call errors to `err`, `ce`, or `call_err`?

Also, what are the next steps for updating the documentation on setters/getters to reflect the changes? Should I start writing a PR to update the docs now, or wait until these changes are approved?

**Comment by dalexeev** [comment]
> So should I name the call errors to `err`, `ce`, or `call_err`?

It seems to me that the name `err` is worse, since it is usually used for `bool` or `Error` codes. For `CallError`, the names `ce`, `call_err`, or the full `call_error` can be used. But in any case, it has the lowest priority. Things like renaming local/internal variables, fixing comment style, adding/removing blank lines are best done on a case-by-case basis, on adjacent lines, as long as it doesn't require a lot of effort and doesn't interfere with the perception of major changes. You are not required to bring the entire file to a single style, there is no point or benefit in this.

> Should I start writing a PR to update the docs now, or wait until these changes are approved?

I don't think this is strictly regulated, but in practice PRs for docs and/or demo projects are usually opened after the PR in the main repository is merged. Since the docs cannot be updated before the main PR is merged. Plus, during the review process, you may be asked to make some changes, which will also require you to update the docs PR.

Therefore, in my opinion, it is better to wait for the final version, so as not to do extra work. It's normal for the documentation (Manual) to be a bit behind the current state, no one expects it to be updated the same day a feature is in master. The exception is the Class Reference, since it's in the main repository, we try to keep it up to date.

**Comment by dalexeev** [comment]
Yes, you can:

```
scons dev_build=true tests=true
bin/<godot executable> --test
```

But this is only for your platform. Tests on GitHub run for different platforms and compilation flags.

**Comment by nlupugla** [comment]
> 

Ah, gotcha. There error was on Linux and I have Windows, so I guess there's not much I can do but hope it passes this time :)

**Comment by AThousandShips** [comment]
This is a new feature and therefore unlikely to be included in 4.1, as we are in feature freeze, so chances are it will have to wait for approval until after the release of 4.1

File: modules/gdscript/gdscript_analyzer.cpp
Comment (AThousandShips) [reply modules/gdscript/gdscript_analyzer.cpp:1303]: I think it might be best shorter, not mentioning the signature but just the issue, unsure about how to word it but this gets pretty long

Comment (nlupugla) [reply modules/gdscript/gdscript_analyzer.cpp:1303]: You're right, the error message could be better.

How about "First argument of a getter "%s" should be a property name and must be a String or StringName"?

File: modules/gdscript/gdscript_analyzer.cpp
Comment (nlupugla) [reply modules/gdscript/gdscript_analyzer.cpp:1329]: Similarly, how about "Second argument of setter %s should be a property name and must be a String or StringName"?

File: modules/gdscript/gdscript_analyzer.cpp
Comment (dalexeev) [reply modules/gdscript/gdscript_analyzer.cpp:1303]: The problem is that 1 argument does **not** have to be a String or StringNamr, generally speaking. It only has to be a string if the mandatory argument count is 2 (and also you can choose not to provide a type hint at all). Otherwise, the first argument of the answer is the "value argument", not the "name argument".

I have used these terms so as not to bring all these details to the user in a short error message, in the assumption that these terms will be used in online documentation.

File: modules/gdscript/gdscript_analyzer.cpp
Comment (nlupugla) [reply modules/gdscript/gdscript_analyzer.cpp:1336]: How about "first argument of setter %s with 2 mandatory arguments should be a property name and must be a String or StringName"

Comment (joao-pedro-braz) [reply modules/gdscript/gdscript_analyzer.cpp:1336]: Yeah that sounds good!

**Comment by dalexeev** [comment]
Need a rebase after #77129.

**Comment by AThousandShips** [comment]
You reset the branch to the base branch, which closes the PR automatically

**Comment by AThousandShips** [comment]
Superseded by: #78991

**Pull request closed without merging**
